crates/dtrace/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+            return Err(Error::OperationError(self.last_error_message()));
+        }
+
+        for option in options.iter() {
+            let name = CString::new(option.0).expect("CString::new failed");
+            let value = CString::new(option.1).expect("CString::new failed");
+            if unsafe { sys::dtrace_setopt(self.inner, name.as_ptr(), value.as_ptr()) } == -1 {
+                return Err(Error::InvalidOption(self.last_error_message()));
+            }
+        }
+
+        if unsafe { sys::dtrace_go(self.inner) } != 0 {
+            return Err(Error::OperationError(self.last_error_message()));
+        }
+
+        Ok(())
+    }
+
+    // Blocks for some time and consumes the probe data collected by DTrace. The |ProgramStatus|
+    // informs about the state of the program execution.
+    //
+    // The duration it blocks depends on `switchrate`, `statusrate` and `aggrate` options used for
+    // the program execution.
+    pub fn wait_and_consume(
+        &mut self,
+        probes: &mut Vec<ProbeData>,
+    ) -> Result<ProgramStatus, Error> {
+        unsafe {
+            sys::dtrace_sleep(self.inner);
+        }
+
+        let user_ptr = &mut *self as *mut _ as *mut std::ffi::c_void;
+        let status = unsafe {
+            sys::dtrace_work(
+                self.inner,
+                std::ptr::null_mut(),
+                Some(probe_callback),
+                Some(probe_action_callback),
+                user_ptr,
+            )
+        };
+
+        probes.append(&mut self.probes);
+
+        match status {
+            sys::dtrace_workstatus_t_DTRACE_WORKSTATUS_OKAY => Ok(ProgramStatus::Ongoing),
+            sys::dtrace_workstatus_t_DTRACE_WORKSTATUS_DONE => Ok(ProgramStatus::Done),
+            _ => Err(Error::OperationError(self.last_error_message())),
+        }
+    }
+
+    /// Instructs the kernel to disable any enabled probe and free the memory.
+    pub fn stop(&self) -> Result<(), Error> {
+        if unsafe { sys::dtrace_stop(self.inner) } == -1 {
+            return Err(Error::OperationError(self.last_error_message()));
+        }
+        Ok(())
+    }
+
+    fn last_error_message(&self) -> String {
+        let message_raw = unsafe { sys::dtrace_errmsg(self.inner, sys::dtrace_errno(self.inner)) };
+        c_char_to_string(message_raw)
+    }
+}
+
+impl Drop for DTrace {
+    fn drop(&mut self) {
+        unsafe {
+            sys::dtrace_close(self.inner);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{DTrace, Error, ProgramStatus};
+
+    #[test]
+    fn dtrace_open_close() -> Result<(), Error> {
+        let mut dtrace = DTrace::new()?;
+
+        dtrace.execute_program(&format!("syscall:::entry {{}}"), &[("bufsize", &"1k")])?;
+
+        let mut probes = Vec::default();
+        assert_eq!(
+            ProgramStatus::Ongoing,
+            dtrace.wait_and_consume(&mut probes)?
+        );
+        assert!(0 < probes.len());
```

Comment:
super nit: I'd reverse the comparison to

```suggestion
        assert!(probes.len() > 0);
```

which IMO reads much closer to natural language "there's *more* than zero probes".

Reply:
Addressed :)